src/sage/categories/finite_fields.py

```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+
+            EXAMPLES::
+                sage: k.<a> = GF((5, 10))
+                sage: k(2).is_square()
+                True
+                sage: k(2)._tonelli()^2 == k(2)
+                True
+                sage: k.quadratic_non_residue()._tonelli()
+                Traceback (most recent call last):
+                ...
+                ValueError: Element is not a square
+            """
+            q = self.parent().cardinality()
+            if not self.is_square():
+                raise ValueError("Element is not a square")
+            g = self.parent().quadratic_non_residue()
+            odd_order = (q - 1).odd_part()
+            even_exp = Integer.valuation(q-1, 2)
+            e = 0
+            for i in range(2, even_exp+1):
+                tmp = self * (pow(g, -e))
+
+                condition = tmp**((q-1)//(2**i)) != self.parent().one()
+                if condition:
+                    e = 2**(i-1) + e
+            h = self * (g**(-e))
+            b = g**(e//2) * h**((odd_order+1)//2)
+            return b
+
+        def _cipolla(self):
+            """
+            Returns a square root of the element if it exists
+            using cipolla's algorithm, more suited if order - 1
+            is highly divisible by 2
+
+            OUTPUT:
+
+            - A square root of the element; raises an error
+              if the element is not a square
+
+            EXAMPLES::
+                sage: k.<a> = GF((5, 10))
+                sage: k(2).is_square()
+                True
+                sage: k(2)._cipolla()^2 == k(2)
+                True
+                sage: k.quadratic_non_residue()._cipolla()
+                Traceback (most recent call last):
+                ...
+                ValueError: Element is not a square
+            """
+            parent = self.parent()
+            q = parent.cardinality()
+            if not self.is_square():
+                raise ValueError("Element is not a square")
+            t = parent.random_element()
+            root = t**2 - 4 * self
+            while root.is_square():
+                t = parent.random_element()
+                root = t**2 - 4 * self
+            from sage.rings.polynomial.polynomial_ring import polygen
+            X = polygen(parent)
+            f = X**2 - t*X + self
+            b = pow(X, (q+1)//2, f)
+            return b
+
+        def sqrt(self, all=False, algorithm='Tonelli'):
+            """
+            Returns the square root of the element if it exists
+
+            INPUT:
+
+            - ``all`` -- boolean (default: ``False``); whether to return a list of
+              all square roots or just a square root
+
+            - ``algorithm`` -- string (default: 'Tonelli'); The algorithm to use
+              among ``'Tonelli'``, ``'Cipolla'``. Most finite fields will have
+              Tonelli to run fast, but in the case where the order minus 1 is
+              highly divisible by 2 Cipolla's method could be faster.
```

Comment:
```suggestion
            - ``algorithm`` -- string (default: 'tonelli'); the algorithm to use
              among ``'tonelli'``, ``'cipolla'``. Tonelli is typically faster but has
              a worse worst-case complexity than Cipolla. In particular, if the
              field cardinality minus 1 is highly divisible by 2 and has a large
              odd factor then Cipolla may perform better.
```

Typically the algorithm options in Sage are in lowercase.

Clearer wording, since elements and the field both have something called an order I think it's more clear to use cardinality here.